mysqld_exporter.go

```diff
@@ -340,6 +357,148 @@ func (e *Exporter) scrape(ch chan<- prometheus.Metric) {
 			)
 		}
 	}
+
+	if *perfTableIOWaitsTime {
+		// Timers here are returned in picoseconds
+		perfSchemaTableWaitsTimeRows, err := db.Query("SELECT OBJECT_SCHEMA, OBJECT_NAME, SUM_TIMER_FETCH, SUM_TIMER_INSERT, SUM_TIMER_UPDATE, SUM_TIMER_DELETE FROM performance_schema.table_io_waits_summary_by_table WHERE OBJECT_SCHEMA NOT IN ('mysql', 'performance_schema')")
+		if err != nil {
+			log.Println("Error running performance schema query on database:", err)
+			e.error.Set(1)
+			return
+		}
+		defer perfSchemaTableWaitsTimeRows.Close()
+
+		var (
+			objectSchema string
+			objectName   string
+			timeFetch    int64
+			timeInsert   int64
+			timeUpdate   int64
+			timeDelete   int64
+		)
+
+		for perfSchemaTableWaitsTimeRows.Next() {
+			if err := perfSchemaTableWaitsTimeRows.Scan(
+				&objectSchema, &objectName, &timeFetch, &timeInsert, &timeUpdate, &timeDelete,
+			); err != nil {
+				log.Println("error getting result set:", err)
+				return
+			}
+			ch <- prometheus.MustNewConstMetric(
+				performanceSchemaTableWaitsTimeDesc, prometheus.CounterValue, float64(timeFetch)/1000000000,
+				objectSchema, objectName, "fetch",
+			)
+			ch <- prometheus.MustNewConstMetric(
+				performanceSchemaTableWaitsTimeDesc, prometheus.CounterValue, float64(timeInsert)/1000000000,
+				objectSchema, objectName, "insert",
+			)
+			ch <- prometheus.MustNewConstMetric(
+				performanceSchemaTableWaitsTimeDesc, prometheus.CounterValue, float64(timeUpdate)/1000000000,
+				objectSchema, objectName, "update",
+			)
+			ch <- prometheus.MustNewConstMetric(
+				performanceSchemaTableWaitsTimeDesc, prometheus.CounterValue, float64(timeDelete)/1000000000,
+				objectSchema, objectName, "delete",
+			)
+		}
+	}
+
+	if *perfIndexIOWaits {
+		perfSchemaIndexWaitsRows, err := db.Query("SELECT OBJECT_SCHEMA, OBJECT_NAME, ifnull(INDEX_NAME, 'NONE') as INDEX_NAME, COUNT_FETCH, COUNT_INSERT, COUNT_UPDATE, COUNT_DELETE FROM performance_schema.table_io_waits_summary_by_index_usage WHERE OBJECT_SCHEMA NOT IN ('mysql', 'performance_schema')")
+		if err != nil {
+			log.Println("Error running performance schema query on database:", err)
+			e.error.Set(1)
+			return
+		}
+		defer perfSchemaIndexWaitsRows.Close()
+
+		var (
+			objectSchema string
+			objectName   string
+			indexName    string
+			countFetch   int64
+			countInsert  int64
+			countUpdate  int64
+			countDelete  int64
+		)
+
+		for perfSchemaIndexWaitsRows.Next() {
+			if err := perfSchemaIndexWaitsRows.Scan(
+				&objectSchema, &objectName, &indexName, &countFetch, &countInsert, &countUpdate, &countDelete,
+			); err != nil {
+				log.Println("error getting result set:", err)
```

Comment:
I assume returning here is fine instead of trying the next row, which will most likely fail in the same way.
However, we have to `e.error.Set(1)` here, right?